proposals/nft.storage-chunked-car-uploads.md

```diff
@@ -0,0 +1,107 @@
+    
+    There is a JS module that will perform the chunking: https://github.com/alanshaw/carbites
+    
+3. Upload the chunks.
+
+    As nft.storage receives the CAR chunks it adds them to IPFS Cluster, pinning the "chunk root" with an **expiry time**.
+
+    Pinning the chunk root node ensures all the blocks in each CAR do not get garbage collected since the chunk root references all the blocks the CAR contains.
+
+    There's no guarantee of order for blocks in a CAR and also no guarantee of order nft.storage will receive the chunks in, so pinning the chunk root mitigates against garbage collection of partial DAGs.
+
+    The expiry time ensures that the chunk root nodes can eventually be garbage collected and also ensures that uploads that never complete can eventually be garbage collected.
+
+4. Pin the root.
+
+    Once all the chunks have been uploaded the actual root CID can be pinned.
+
+    The `https://api.nft.storage/pins` API can perform this action (which is an existing part of the pinning service API).
+
+    We may choose to create a separate HTTP endpoint for this.
+    
+    This could be done first, but it would cause IPFS to do work to attempt to find the blocks before they are uploaded.
+    
+    Pinning the root last also ensures nft.storage has all the data for the NFT before it is registered to the user. It's an insentive to continue uploading all the chunks, avoiding any blame for losing data in the case where a user ceases uploading before completion.
+    
+### Alternative
+
+For completeness, an alternative was suggested that could also work:
+
+1. Generate a CAR file for the content.
+2. Split the CAR file into chunks of <100MB. Chunks following the first have an ["empty" CID `bafkqaaa`](https://cid.ipfs.io/#bafkqaaa) as their root node. [See recommendation](https://github.com/ipld/specs/blob/master/block-layer/content-addressable-archives.md#number-of-roots).
+3. Upload and pin the root. We'd need to ensure the first CAR contains the root node.
+4. Upload the remaining chunks.
+
+Advantages:
+
+* Simplified chunked CAR file creation.
+* No separate pin request.
+
+Disadvantages:
+
+* An NFT is registered to the user before all data is received.
+* An out of order CAR chunk _may_ be garbage collected since it could be unreachable from the pinned root node.
+* Failure to upload all chunks will likely result in an NFT registered to a user with missing data forever.
+* A CAR file created with blocks in random order could get garbage collected.
+
+## Success/acceptance criteria (optional)
+_How do we know we're done with this project? How do we know we're successful? This field is OPTIONAL for the first draft of an MPP. Sometimes this field needs to be filled out once we have more detail on the shape of the actual solution._
+
+* Assets larger than 100MB can be uploaded to nft.storage.
+* A user uploads a 1GB asset to nft.storage.
+
```

Comment:
```suggestion
* Upload endpoint is part of [ipfs/pinning-services-api-spec](https://github.com/ipfs/pinning-services-api-spec/)

```
Could this proposal be an opportunity to come up with a spec for a generic endpoint that any public pinning service could provide? Feels like something we discussed in https://github.com/ipfs/pinning-services-api-spec/issues/7 

My main drive here is to remove vendor lock-in and make all the CAR-splitting tooling usable outside nft.storage (removing barriers if someone chooses to run their own infra).

I was tempted to add "ipfs-cluster  can be used for self-hosting", but I am not sure if nft.storage uses it directly. Perhaps ["making ipfs-cluster compatible with pinning service api"](https://github.com/ipfs/ipfs-cluster/issues/1213) should be a separate project on its own?

Reply:
I think you meant ipfs/pinning-services-api-spec#73, right?

> My main drive here is to remove vendor lock-in and make all the CAR-splitting tooling usable outside nft.storage (removing barriers if someone chooses to run their own infra).

Do you mean supporting single car file uploads via the pinning service API, or this multi car file approach? Note that the atomicity properties which were the subject of some debate in that issue are not really present in this multi-car approach which makes things pretty easy (although we could still push on the atomic upload + pin of a single car file). On the other hand, the multi-car approach taken here seems pretty specific to some assumptions about nft.storage which makes asking other people to implement it seem like a stretch.

IIUC some assumptions in this proposal which are coupled to nft.storage are:
- The expiry time ensures that the chunk root nodes can eventually be garbage collected and also ensures that uploads that never complete can eventually be garbage collected.
    - Implies the application can properly track + label pins for the same chunk root even by the same user (e.g. uploading overlapping data twice)
    - Implies the application can support time-expired pins and garbage collection of them
- Pinning the root last also ensures nft.storage has all the data for the NFT before it is registered to the user. It's an insentive to continue uploading all the chunks, avoiding any blame for losing data in the case where a user ceases uploading before completion.
     - Implies the application is unable/unwilling to inform the user of whether their upload is in progress or not (e.g. when I upload a 50MB file to GitHub I can see the progress report of how far along it is)

Reply:
I'll reply to these comments properly tomorrow but just quickly  - IPFS Cluster provides the expiring pin functionality.